core/src/test/java/io/grpc/internal/SerializingExecutorTest.java

```diff
@@ -0,0 +1,301 @@
+        callOrder.add(op);
+      }
+    }
+
+    serializingExecutor.execute(new FakeOp(1));
+    serializingExecutor.execute(new FakeOp(2));
+    fakePool.runAll();
+
+    assertEquals(ImmutableList.of(1, 2), callOrder);
+  }
+
+  @Test
+  public void testRuntimeException_doesNotStopExecution() {
+
+    final AtomicInteger numCalls = new AtomicInteger();
+
+    Runnable runMe = new Runnable() {
+      @Override
+      public void run() {
+        numCalls.incrementAndGet();
+        throw new RuntimeException("FAKE EXCEPTION!");
+      }
+    };
+
+    serializingExecutor.execute(runMe);
+    serializingExecutor.execute(runMe);
+    fakePool.runAll();
+
+    assertEquals(2, numCalls.get());
+  }
+
+  @Test
+  public void testInterrupt_doesNotStopExecution() {
+
+    final AtomicInteger numCalls = new AtomicInteger();
+
+    Runnable runMe = new Runnable() {
+      @Override
+      public void run() {
+        numCalls.incrementAndGet();
+      }
+    };
+
+    Thread.currentThread().interrupt();
+
+    serializingExecutor.execute(runMe);
+    serializingExecutor.execute(runMe);
+    fakePool.runAll();
+
+    assertEquals(2, numCalls.get());
+    assertTrue(Thread.interrupted());
+  }
+
+  @Test
+  public void testDelegateRejection() {
+    final AtomicInteger numCalls = new AtomicInteger();
+    final AtomicBoolean reject = new AtomicBoolean(true);
+    final SerializingExecutor executor = new SerializingExecutor(
+        new Executor() {
+          @Override public void execute(Runnable r) {
+            if (reject.get()) {
+              throw new RejectedExecutionException();
+            }
+            r.run();
+          }
+        });
+    Runnable task = new Runnable() {
+      @Override
+      public void run() {
+        numCalls.incrementAndGet();
+      }
+    };
+    try {
+      executor.execute(task);
+      fail();
+    } catch (RejectedExecutionException expected) {
+      // No-op
+    }
+    assertEquals(0, numCalls.get());
+    reject.set(false);
+    executor.execute(task);
+    assertEquals(2, numCalls.get());
+  }
+
+  @Test
+  public void testTaskThrowsError() throws Exception {
+    class MyError extends Error {}
+
+    final CyclicBarrier barrier = new CyclicBarrier(2);
+    // we need to make sure the error gets thrown on a different thread.
+    ExecutorService service = Executors.newSingleThreadExecutor();
+    try {
+      final SerializingExecutor executor = new SerializingExecutor(service);
+      Runnable errorTask = new Runnable() {
+        @Override
+        public void run() {
+          throw new MyError();
+        }
+      };
+      Runnable barrierTask = new Runnable() {
+        @Override
+        public void run() {
+          try {
+            barrier.await();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+      };
+      executor.execute(errorTask);
+      service.execute(barrierTask);  // submit directly to the service
```

Comment:
It guarantees that the barrier task has been executed one after the executor has observed the error and stopped executing (see earlier note about allowing error to pass up). When the barrier task is then scheduled onto the serializing executor it should cause the queue to start being processed again.